Update DataContexts API to remove references to Deferred

markfields opened the pull request:
The use of Deferred is an implementation detail. This also improves the API when binding, with a very minor and I believe inconsequential reordering of some statements.

Next small PR I have in mind (for Monday) is to keep the not bound contexts separate from the all good ones so we can get rid of the `as LocalFluidDataStoreContext` cast.

markfields committed:
Get rid of deferred.isCompleted

markfields committed:
Adjust public API to avoid Deferred

markfields committed:
Clean up redundant id param

markfields committed:
Clean up API around binding

Note: moves notBoundContexts.delete relative to other code in
DataStores.bindFluidDataStore

anthony-murphy commented on packages/runtime/container-runtime/src/dataStoreContexts.ts:
```diff
@@ -106,33 +109,40 @@ import { FluidDataStoreContext, LocalFluidDataStoreContext } from "./dataStoreCo
+     * Update this context as bound
+     */
+    public bind(id: string) {
+        this.notBoundContexts.delete(id);
```
we should assert that there is a notBoundContext removed

anthony-murphy commented on packages/runtime/container-runtime/src/dataStoreContexts.ts:
```diff
@@ -68,13 +68,12 @@ import { FluidDataStoreContext, LocalFluidDataStoreContext } from "./dataStoreCo
      */
-    public prepContextForBind(id: string): LocalFluidDataStoreContext {
+    public getUnbound(id: string): LocalFluidDataStoreContext {
         assert(this.notBoundContexts.has(id), "Store being bound should be in not bounded set");
```
these assert messages should be updated

anthony-murphy reviewed (approved):
:shipit:

anthony-murphy commented on packages/runtime/container-runtime/src/dataStores.ts:
```diff
@@ -270,13 +269,16 @@ export class DataStores implements IDisposable {
     public async getDataStore(id: string, wait: boolean): Promise<IFluidDataStoreChannel> {
-        const deferredContext = this.contexts.prepDeferredContext(id);
+        const existingContext = this.contexts.get(id);
+        if (existingContext !== undefined && !this.contexts.isNotBound(id)) {
```
this method has always been subtle. it looks right to me, but unit tests would be great if they don't already exist. it might be easier to test dataStoreContexts, and since this is the only place we use waitForContext it might make sense to move this logic to dataStoreContexts

markfields replied:
@anthony-murphy - Great point, I'm thinking this method (minus the call to `.realize()`) moves to `DataStoreContexts.getAttached`. Or if `Attached` is not the right adjective for the context for a datastore that exists on the server, what is? `Remoted`? I'm reading through `DataStoreContext.ts` now trying to understand the difference between the different concrete classes...

anthony-murphy replied:
The semantics are complicated, and it doesn't fit neatly with any of our current verb, as when detached container or local it only returns bound contexts, not unbound. But for an attached container if can also be used for remote context. Personally, i would avoid referencing in either. I like just calling it waitForContext as it is now, but just putting more implementation in it, and then adding a bunch of comments that describe the behavior

markfields replied:
Ok gotcha.  Will update.

markfields replied:
Went with `getBoundOrRemoted` to match `addBoundOrRemoted` (note the added `d` to `Remote` to match the class `RemotedDataStoreContext)

markfields committed:
PR feedback

Also replaced a few asserts with returning undefined for caller to assert about.

markfields committed:
Consistent naming

markfields committed:
Slightly less coupled assert in test

markfields committed:
getBoundOrRemoted - Clearer API around waiting or not

anthony-murphy commented on packages/runtime/container-runtime/src/dataStoreContexts.ts:
```diff
@@ -68,13 +68,15 @@ import { FluidDataStoreContext, LocalFluidDataStoreContext } from "./dataStoreCo
-        assert(this._contexts.has(id), "Attempting to bind to a context that hasn't been added yet");
+    public getUnbound(id: string): LocalFluidDataStoreContext | undefined {
+        const context = this._contexts.get(id);
+        if (!context || !this.notBoundContexts.has(id)) {
```
```suggestion
        if (context === undefined || !this.notBoundContexts.has(id)) {
```
thought we had an eslint rule against truthy checks. might be off for this project

anthony-murphy commented on packages/runtime/container-runtime/src/dataStores.ts:
```diff
@@ -270,14 +271,13 @@ export class DataStores implements IDisposable {
 
     public async getDataStore(id: string, wait: boolean): Promise<IFluidDataStoreChannel> {
-        const deferredContext = this.contexts.prepDeferredContext(id);
+        const contextP = this.contexts.getBoundOrRemoted(id, wait);
```
why not just await here, rather than hanging on to the promise

markfields replied:
Well I did before but then it looked weird, as if we're ignoring the `wait` param. See the last commit. Happy to revert if you like it better before, I was on the fence.

anthony-murphy replied:
the method is always async, regardless of the wait param, so the caller will always need to await. i'm not sure i see what you are trying to do.

anthony-murphy commented on packages/test/end-to-end-tests/src/test/loaderTest.spec.ts:
```diff
@@ -157,7 +157,8 @@ describe("Loader.request", () => {
             assert(false, "Loader pause flags doesn't pause container op processing");
         } catch (e) {
-            assert.strictEqual(e.message, `DataStore ${newDataStore.id} does not exist`);
+            const topFrame: string | undefined = e?.stack.split("\n")[1].trimLeft();
```
why did this have to change?

markfields replied:
Assert message changed.  I thought this was a slightly looser coupling.

anthony-murphy commented on packages/runtime/container-runtime/src/dataStoreContexts.ts:
```diff
@@ -88,15 +90,27 @@ import { FluidDataStoreContext, LocalFluidDataStoreContext } from "./dataStoreCo
+     * @param wait If false, return undefined if the context isn't present and ready now. Otherwise return the Promise.
      */
-    public prepDeferredContext(id: string): Deferred<FluidDataStoreContext> {
+    public getBoundOrRemoted(id: string, wait: boolean): Promise<FluidDataStoreContext> | undefined {
```
shouldn't this need to be async?

markfields replied:
It doesn't always return a promise (sometimes straight `undefined`). See other comment thread about the last commit's refactor here.

anthony-murphy replied:
i would put the undefined union into to promise, as it makes the api easier to use. It's not free, but it won't cause a perf issue here, it's only a problem if don't many many times on a critical path

markfields committed:
Remove truthy check

Co-authored-by: Tony Murphy <anthony.murphy@microsoft.com>

markfields committed:
Revert "getBoundOrRemoted - Clearer API around waiting or not"

This reverts commit 6883e5491c3d75d594b0d90eccc8e400b5189b84.
